internal/mode/static/state/graph/httproute.go

```diff
@@ -94,6 +99,101 @@ func buildRoutesForGateways(
 	return routes
 }
 
+func buildMirrorRoutesForGateways(
+	validator validation.HTTPFieldsValidator,
+	httpRoutes map[types.NamespacedName]*v1.HTTPRoute,
+	routes map[types.NamespacedName]*Route,
+	gatewayNsNames []types.NamespacedName,
+) map[types.NamespacedName]*Route {
+	var mirroredRoutesAndRoutes map[types.NamespacedName]*Route
+	if len(routes) == 0 {
+		mirroredRoutesAndRoutes = make(map[types.NamespacedName]*Route)
+	} else {
+		mirroredRoutesAndRoutes = routes
+	}
+
+	if len(gatewayNsNames) == 0 {
+		return nil
+	}
+
+	for _, ghr := range httpRoutes {
+		routeRules := ghr.Spec.Rules
+		for i, rule := range routeRules {
+			filters := rule.Filters
+			for _, filter := range filters {
+				if filter.RequestMirror == nil {
+					continue
+				}
+				matchesPath := rule.Matches[i].Path.Value
+				mirrorGhr := createMirrorRoute(matchesPath, ghr, filters, filter)
+				r := buildRoute(validator, mirrorGhr, gatewayNsNames)
+				if r != nil {
+					mirroredRoutesAndRoutes[client.ObjectKeyFromObject(mirrorGhr)] = r
+				}
+			}
+		}
+	}
+
+	return mirroredRoutesAndRoutes
+}
+
+func createMirrorRoute(
+	matchesPath *string,
+	ghr *v1.HTTPRoute,
+	filters []v1.HTTPRouteFilter,
+	filter v1.HTTPRouteFilter,
+) *v1.HTTPRoute {
+	mirrorGhr := &v1.HTTPRoute{
+		ObjectMeta: createMirrorObjectMeta(ghr),
+		Spec: v1.HTTPRouteSpec{
+			CommonRouteSpec: v1.CommonRouteSpec{
+				ParentRefs: ghr.Spec.ParentRefs,
+			},
+			Hostnames: ghr.Spec.Hostnames,
+			Rules: []v1.HTTPRouteRule{
+				buildMirrorRouteRule(matchesPath, filters, filter),
+			},
+		},
+	}
+	return mirrorGhr
+}
+
+func buildMirrorRouteRule(
+	matchesPath *string,
+	filters []v1.HTTPRouteFilter,
+	filter v1.HTTPRouteFilter,
+) v1.HTTPRouteRule {
+	filterBackendRef := filter.RequestMirror.BackendRef
+	return v1.HTTPRouteRule{
+		Matches: []v1.HTTPRouteMatch{
+			{
+				Path: &v1.HTTPPathMatch{
+					Type:  helpers.GetPointer(v1.PathMatchExact),
+					Value: helpers.CreateMirrorPathWithBackendRef(matchesPath, filterBackendRef),
+				},
+			},
+		},
+		Filters: removeMirrorFilters(filters),
+		BackendRefs: []v1.HTTPBackendRef{
+			{BackendRef: createMirrorBackendRef(filter)},
+		},
+	}
+}
+
+func createMirrorObjectMeta(ghr *v1.HTTPRoute) metav1.ObjectMeta {
+	objectMeta := ghr.ObjectMeta.DeepCopy()
+	objectMeta.SetName(fmt.Sprintf("%s-mirror", ghr.ObjectMeta.GetName()))
+	return *objectMeta
+}
+
+func createMirrorBackendRef(filter v1.HTTPRouteFilter) v1.BackendRef {
+	backendRef := filter.RequestMirror.BackendRef
+	mirrorBEF := v1.BackendRef{
+		BackendObjectReference: backendRef,
+	}
+	return mirrorBEF
+}
```

Comment:
Just so I understand properly, you're essentially creating an additional location block for the backend that's referred to in the mirror filter, and the main location block for the path will mirror to that block, which contains all the same settings as the route, except for the mirror filter?

If that's accurate, I wonder if this new location should be a named "internal" location, similar to how we do with our [http match redirect locations](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/internal/mode/static/nginx/config/servers.go#L221). Also, if we mirror the traffic to this new location with a custom path, does that custom path get added to the request to the mirrored backend because of `$request_uri` in the `proxy_pass`? Or does the `mirror` directive strip out the custom path and only preserve the real path?

Reply:
> Just so I understand properly, you're essentially creating an additional location block for the backend that's referred to in the mirror filter, and the main location block for the path will mirror to that block, which contains all the same settings as the route, except for the mirror filter?
>
Exactly
 
> If that's accurate, I wonder if this new location should be a named "internal" location, similar to how we do with our [http match redirect locations](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/internal/mode/static/nginx/config/servers.go#L221). Also, if we mirror the traffic to this new location with a custom path, does that custom path get added to the request to the mirrored backend because of `$request_uri` in the `proxy_pass`? Or does the `mirror` directive strip out the custom path and only preserve the real path?
>
Again, bingo. It preserves the real path

Reply:
Good to know, thanks. I wonder if we can still use a [named location](https://nginx.org/en/docs/http/ngx_http_core_module.html#location) (prefixed with `@`) to keep it as an internal-only location, or if this would break the mirror functionality. One reason for this is to avoid potential path collision with an HTTPRoute that *could* be configured with the custom path we generate.

Reply:
Named locations aren't supported with the mirror directive 😞

Reply:
Ah, too bad